merge.go

```diff
@@ -0,0 +1,102 @@
+package zenodb
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/getlantern/bytemap"
+	"github.com/getlantern/errors"
+	"github.com/getlantern/goexpr"
+	"github.com/getlantern/wal"
+	"github.com/getlantern/zenodb/encoding"
+	"github.com/getlantern/zenodb/sql"
+)
+
+var (
+	emptyOffset = make(wal.Offset, wal.OffsetSize)
+)
+
+// FilterAndMerge merges the specified inFiles into the given outFile, where
+// inFiles and outFiles are all valid filestore files. The schema is based on
+// the named table. If whereClause is specified, rows are filtered by comparing
+// them to the whereClause. The merge is performed as a disk-based merge in
+// order to use minimal memory. If shouldSort is true, the output will be sorted
+// by key.
+func (db *DB) FilterAndMerge(table string, whereClause string, shouldSort bool, outFile string, inFiles ...string) error {
+	t := db.getTable(table)
+	if t == nil {
+		return errors.New("Table %v not found", table)
+	}
+	return t.filterAndMerge(whereClause, shouldSort, outFile, inFiles)
+}
+
+func (t *table) filterAndMerge(whereClause string, shouldSort bool, outFile string, inFiles []string) error {
+	okayToReuseBuffers := false
+	rawOkay := false
+
+	filter, err := whereFor(whereClause)
+	if err != nil {
+		return err
+	}
+
+	// Find highest offset amongst all infiles
+	var offset wal.Offset
+	for _, inFile := range inFiles {
+		nextOffset, _, offsetErr := readWALOffset(inFile)
+		if offsetErr != nil {
+			return errors.New("Unable to read WAL offset from %v: %v", offsetErr)
+		}
+		if nextOffset.After(offset) {
+			offset = nextOffset
+		}
+	}
+
+	// Create output file
+	out, err := os.OpenFile(outFile, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0644)
+	if err != nil {
+		return errors.New("Unable to create outFile at %v: %v", outFile, err)
+	}
+	defer out.Close()
+
+	fso := &fileStore{
+		t:      t,
+		fields: t.fields,
+	}
+	cout, err := fso.createOutWriter(out, t.fields, offset, shouldSort)
```

Comment:
Zeno works using a write-ahead-log (WAL). The leader (ingestion) node writes incoming data into the WAL, and the follower nodes tail the WAL in order to get updates. These are first merged into an in-memory datastructure (memstore) and then periodically committed to disk into a filestore. At this point, we commit not just the in-memory data, but also the WAL offset for the latest ingested data that contributed to the memstore. This way, if the follower goes down, it knows where to start tailing the WAL when it starts back up.

When merging multiple files, we have to decide what WAL offset represents that data. Right now, I just use the highest WAL offset from any of the input files. This is definitely not the only way, and it may not even be a good way, but I had to pick something :)